spec/request/session/create_spec.rb

```diff
@@ -0,0 +1,59 @@
+require 'rails_helper'
+
+RSpec.describe 'POST #create sign_in', type: :request do
+  let!(:user_attrs) { { user: FactoryBot.attributes_for(:user) } }
+
+  subject do
+    post '/api/v1/users/sign_in', params: {
+      email: user_attrs[:user][:email], password: user_attrs[:user][:password]
+    }, as: :json
+  end
+
+  context 'when valid' do
+    context 'when registered and confirmed' do
+      let!(:user) { FactoryBot.create(:confirmed_user) }
+
+      it 'does get a sucessful answer' do
+        subject
+        expect(response).to be_successful
+      end
+
+      it 'does create a new session' do
+        expect { subject }.to(change { User.last.current_sign_in_at })
+      end
+    end
+  end
+
+  context 'when invalid' do
+    context 'when registered and not confirmed' do
+      let!(:user) { FactoryBot.create(:user) }
+
+      it 'does not get a sucessful answer' do
+        subject
+        expect(response).not_to be_successful
+      end
+      it 'does not sign_in' do
+        subject
+        expect { subject }.not_to(change { User.last.current_sign_in_at })
+      end
+    end
+
+    context 'when registered and confirmed but invalid credentials' do
+      let!(:user) { FactoryBot.create(:confirmed_user) }
+
+      it 'does not get a sucessful answer' do
+        post '/api/v1/users/sign_in', params: {
+          email: user_attrs[:user][:email], password: 'another'
+        }, as: :json
+        expect(response).not_to be_successful
+      end
+
+      it 'does not sign_in' do
+        post '/api/v1/users/sign_in', params: {
+          email: user_attrs[:user][:email], password: 'another'
```

Comment:
Something you can do here:
In the subject above
```ruby
subject do
    post new_path, params: { email: user_attrs[:user][:email], password: password }, as: :json
end
```
Then you can redefine the `let` inside each context, and the subject will use that let.

```ruby
context 'when password not valid'
  let(:password) {  'invalid' }

  it .... do
    subject # will use subject with password = 'invalid'
    expect...
  end
end

context 'when password is a number'
  let(:password) {  123 }

  it .... do
    subject # will use subject with password = 123
    expect...
  end
end
```